### ISSUE TITLE
Not handling all combinations of named enum values and booleans should give a CS8509 warning regardless of order, even if warnings are disabled for CS8524

### ISSUE THREAD

**Issue by olatoft** [dotnet/roslyn#64698]
[Summary(jcouv):] The ask is to report exhaustiveness problems using named enum values (over unnamed ones) where possible.

**Version Used**:
dotnet version 6.0.110
Microsoft (R) Build Engine version 17.0.1+b177f8fa7 for .NET

**Steps to Reproduce**:

A CS8524 warning is given if we forget to handle an unnamed enum value, for instance `(Values)2`. I want to disable warnings for CS8524, in order to get a warning, namely CS8509, if and only if I forget to handle a named enum value, or a combination of named enum values and booleans.

```csproj
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net6.0</TargetFramework>
    <RootNamespace>test_exhaustive</RootNamespace>
    <ImplicitUsings>enable</ImplicitUsings>
    <Nullable>enable</Nullable>
    <NoWarn>CS8524</NoWarn>
  </PropertyGroup>

</Project>
```

The following, with `Value1` at the top, gives "warning CS8509: The switch expression does not handle all possible values of its input type (it is not exhaustive). For example, the pattern '(Values.Value1, false)' is not covered.", as I expect:

```c#
var test1 = (Values.Value0, true) switch
{
    (Values.Value1, true) => true,
    (Values.Value0, true) => true,
    (Values.Value0, false) => true,
};

enum Values
{
    Value0,
    Value1
}
```

The following, with the boolean first, also gives the same warning, as I expect:

```c#
var test2 = (true, Values.Value0) switch
{
    (true, Values.Value0) => true,
    (false, Values.Value0) => true,
    (true, Values.Value1) => true,
};
```

However, the following, with `Value1` at the bottom and the boolean last, gives no warnings, which I do not expect:

```c#
var test3 = (Values.Value0, true) switch
{
    (Values.Value0, true) => true,
    (Values.Value0, false) => true,
    (Values.Value1, true) => true,
};
```

If I do handle some unnamed enum values without handling `(Values.Value1, false)`, regardless of whether warnings for CS8524 are disabled or not, I do get the same warning as in `test1` and `test2`, as I expect:

```c#
var test4 = (Values.Value0, true) switch
{
    (Values.Value0, true) => true,
    (Values.Value0, false) => true,
    (Values.Value1, true) => true,
    (_, true) => true,
};
```

**Expected Behavior**:
A CS8509 warning telling me that I forgot to handle the pattern `(Values.Value1, false)` for all of the above examples, regardless of order.

**Actual Behavior**:
No warning for `test3`, even if not all combinations of named enum values and booleans are handled.

**Comment by olatoft** [comment]
Thanks for looking into the issue :) I'm sad to see it closed by design, because this is a feature I would love to have. However, I understand your arguments about complexity and being careful about adding strong guarantees.

Is there any other way in C# to get a warning, or ideally an error, if I forget to handle all cases of a discriminated union and combinations of discriminated unions? From what I now know, it doesn't seem to be possible with named enum values and booleans, unless I manually make sure to put them in the right order, and it doesn't seem to be possible with subclasses, as described in #64497. Would it be possible to implement such a feature in Roslyn? Or would it require a change in the C# language design? Or is it simply not realistic to expect such a feature in C#?

**Comment by jcouv** [comment]
We are actively discussing some design options around closed type hierarchies and discriminated unions. One of the proposals is to add a `[Closed]` on a type hierarchy or enum, which would inform `switch` and patterns that the set of possible values is restricted.
Working group notes are posted about every other week in https://github.com/dotnet/csharplang/tree/main/meetings/working-groups/discriminated-unions

**Comment by olatoft** [comment]
That's exciting to hear! Thanks for the info, and good luck with the work :)
